### PR TITLE
Add logger and its tests

### PR DESCRIPTION
**Pull request by AnkshitJain**

### CODE REVIEW

**Comment by prasadtalasila** [comment]
@AnkshitJain Please check the codeclimate issues. Some of the issues seem related to the inability of codeclimate to understand unit tests. There might be a general solution to this oft encountered problem. Please look into it.

The code looks fine. Your asserts in unit tests are in BDD format. Please check the examples of chai to make sure that we are following good assertion practices. If you can provide the URL / documentation page from which you adopted the assertion style, that would help me understand the motivation better.

**Comment by AnkshitJain** [comment]
The codeclimate error related to unit tests have been fixed. The eslintrc has been updated to include the mocha environment.

**Comment by AnkshitJain** [comment]
@prasadtalasila There is a lot of duplication in code. I have tried to remove a significant amount of them by including two helper functions testTwoLogFiles and testSingleLogFile. Please provide a design suggestion for removing the existing errors.

**Comment by AnkshitJain** [comment]
Documentation for the unit tests methods can be found [here](http://chaijs.com/api/bdd/). The style guide is specified [here](http://chaijs.com/guide/styles/). It gives an overview of the three different styles available in chai.

**Comment by prasadtalasila** [comment]
@AnkshitJain I am fine with should style of assertions as long as we can do the same for unit / functional tests on components too. The style guide seems to indicate as much. It would be better to write a minimal example to confirm that fact for yourself.

Let me check the Logger.js code and revert back with suggestions.

**Comment by AnkshitJain** [comment]
@prasadtalasila I have updated the commit and removed some codeclimate errors. Please have a look at it.

**Comment by prasadtalasila** [comment]
@AnkshitJain I don't see any deploy/configs/logger/logger.json. Is there any reason to omit the config file?

**Comment by AnkshitJain** [comment]
The json file has not been included. I have made log files to be stored as simple files. The max file length can be included in the json file, but it isn't done in this PR. If the existing implementation is acceptable, and it's tests fulfill the expected purpose, we can go forward with the json file.

**Comment by prasadtalasila** [comment]
@AnkshitJain Here are a few pointers to clean up the code.

1. We don't need to create multiple winston objects to log messages to different files. Winston, by default, supports different transports (files) for different log levels. Please [example](https://github.com/winstonjs/winston#multiple-transports-of-the-same-type).
1. We need to support different log levels via configuration file. The log level can be used to create winston logger of required log level. This can be done inside the constructor. It is also better to make the winston logger object private. Please see examples [1](https://stackoverflow.com/questions/22156326/private-properties-in-javascript-es6-classes), [2](https://stackoverflow.com/questions/13833204/how-to-set-a-js-object-property-name-from-a-variable) and [3](http://exploringjs.com/es6/ch_classes.html).
1. We can provide the same methods as Winston (for example, logger.log() and logger.info() etc). In a way, we are just restricting the Winston logger object to a pre-configured object that can't be configured on demand.

These ideas should dramatically simplify logger.js.

BTW, I am not sure if the file name should be either logger.js or Logger.js. Please choose a prevailing naming convention.

**Comment by AnkshitJain** [comment]
We can create a single winston object and pass multiple transport files to it. However, I can explain with an example what will happen in that case:     
Suppose we have a single winston logger object that logs to three different files pertaining to three levels: error, info and debug. Note according to priority levels, error > info > debug.

Now, when we log an `debug` message, say e1, e1 goes to the file related to `debugLog`. This is the required and expected behavior. 

However, when we log an `info` message, say e2,  e2 goes to the `infoLog` as well as `debugLog`. In fact, if we included verbose and silly, the info log would be logged there as well. 

Same thing happens with `error` logs. They will be logged in `infoLog` and `debugLog`  along with `errorLog`. This is not required and just duplication of log messages across various files.

**Comment by AnkshitJain** [comment]
Regarding the second point, the logger objects that are `logError`, `logInfo`, and `logDebug`; are all class private variables. The methods `error()`, `info()`, and `debug()` are just wrappers around the actual logging methods for the appropriate logger object.     
Also do you suggest we specify the log levels in the config and determine at runtime what logger objects are available for use?

**Comment by AnkshitJain** [comment]
For the third point, since the logger objects are private, they can't be accessed unless you use external libraries like [rewire](https://www.npmjs.com/package/rewire), which are only used in tests to test the behaviour of module private methods.

**Comment by AnkshitJain** [comment]
The filename convention followed is lower case letters with underscores in all other js files. Also as specified [here](https://google.github.io/styleguide/jsguide.html#source-file-basics), I think `logger.js` should be preferred over `Logger.js`.

**Comment by prasadtalasila** [comment]
A code gist from logger.js module is shown below.
 
```javascript
class Logger {
  constructor(component) {
      ...
      this.logError = createLoggerObject(this.logDirectory, this.componentRole, 'error');
      ...
    }
  }

  error(log, response) {
    this.logError.log('error', log, () => {
      response(true);
    });
  }
...
}
```
The constructor accesses present object via `this.`, thus making the property `logError` available to anyone who wants to access the same. We can either put an underscore infront of the property name as a convention for private variable or use closure in constructor to enforce private property.


Another point is related to restricting the log-levels based on configuration. If the log-level is specified to `warning`, then all log messages at `info` level are silently ignored. The `info` or `debug` log messages should not generate run-time errors. One way to do this would be to
```javascript
    switch (level) {
        case 'error':
            this.logError = createLoggerObject(this.logDirectory, this.componentRole, 'error');
            break;
        case 'warn':
            this.logError = createLoggerObject(this.logDirectory, this.componentRole, 'error');
            this.logWarn = createLoggerObject(this.logDirectory, this.componentRole, 'warn');
            break;
        ...
        default:
            break;
    }

      error(log) {
          if (this.logError != undefined)
              this.logError.log('error', log)
     }
```
I am not really convinced of the need for having a callback function to the logger. The logger functions are supposed to be stand alone statements; we don't put logger statements in any assignment statements making the return values and callbacks unnecessary.

**Comment by AnkshitJain** [comment]
The idea of having an environment variable containing the logger.json path isn't an accepted convention.[1](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-dynamic-require.md), [2](https://github.com/benmosher/eslint-plugin-import/issues/567)
With our case, it would be acceptable if we ignore that rule for now.

**Comment by AnkshitJain** [comment]
@prasadtalasila Please have a look at the updated commit.

File: util/test/loggerTest.js [hunk 1]
```diff
@@ -0,0 +1,268 @@
+/* eslint no-underscore-dangle: [2, { "allow": ["__get__", "_logInfo", "_logError",
+"_logDebug"] }] */
+
+/* The writing in log file is async. Hence, even after specifying a wait period,
+there is a chance the tests may fail. */
+
+const chai = require('chai');
+const winston = require('winston');
+const Logger = require('../logger.js').Logger;
+const { exec } = require('child_process');
```
Comment (prasadtalasila) [inline util/test/loggerTest.js]: why is this a different import convention?

File: util/test/loggerTest.js [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+const logDirectory = config.logDirectory;
+const createLoggerObject = rewire('../logger.js').__get__('createLoggerObject');
+
+chai.should();
+chai.use(chaiFs);
+chai.use(dirtyChai);
+
+/* These variables will be used in the following tests.
+They have been declared in the global space to avoid repetition. */
+let logger;
```
Comment (prasadtalasila) [inline util/test/loggerTest.js]: The names can be MSLogger and LBLogger. Similarly, the logger config objects can be named as MSLoggerConfig and LBLoggerConfig. Such self-descriptive names make the code more readable.

File: util/test/loggerTest.js [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+  port: '8081',
+  cmd: 'log',
+};
+const secondComponentConfig = {
+  role: 'main_server',
+  hostname: 'localhost',
+  port: '9000',
+  cmd: 'log',
+};
+const logMessage = ['Error message', 'Debug message', 'Info message'];
```
Comment (prasadtalasila) [inline util/test/loggerTest.js]: logMessages and logLevels are syntactically correct names here.

File: util/test/loggerTest.js [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+
+const fileContentChecker = function fileContentChecker(fileName, message, level, response) {
+  const filePath = path.join(logDirectory, fileName);
+  const fileContent = fs.readFileSync(filePath, 'utf8');
+  (JSON.parse(fileContent).message).should.equal(message);
+  (JSON.parse(fileContent).level).should.equal(level);
+  response();
+};
+
+describe('Test helper functions', () => {
```
Comment (prasadtalasila) [inline util/test/loggerTest.js]: The describe statements need to be positive and definitive. For example, here, we can use, "Helper functions work correctly"

File: util/test/loggerTest.js [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+  it('Should delete log directory', (done) => {
```
Comment (prasadtalasila) [inline util/test/loggerTest.js]: "Deletes log directory"

File: util/test/loggerTest.js [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+    fs.mkdirSync(logDirectory);
+    logDirectory.should.be.a.directory().and.empty();
+    deleteLogDirectory(() => {
+      logDirectory.should.not.be.a.path();
+      done();
+    });
+  });
+
+  it('Should check contents of a file', (done) => {
```
Comment (prasadtalasila) [inline util/test/loggerTest.js]: Logger file gets written properly.

File: util/test/loggerTest.js [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+    fs.writeFileSync(logPath[0], JSON.stringify(writeData), 'utf8');
+    fileContentChecker(log[0], logMessage[0], logLevel[0], () => {
+      deleteLogDirectory(() => {
+        done();
+      });
+    });
+  });
+});
+
+describe('Creation of a logger object', () => {
```
Comment (prasadtalasila) [inline util/test/loggerTest.js]: Creates a logger object

Comment (prasadtalasila) [reply util/test/loggerTest.js]: please make similar changes to the following test statements.

File: util/logger.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
+/* eslint no-underscore-dangle: [2, { "allow": ["_logInfo", "_logError", "_logDebug"] }] */
+/* eslint import/no-dynamic-require: 0 */
+const winston = require('winston');
+const fs = require('fs');
+const path = require('path');
+/* The environment variable LOGGERCONFIG will contain the path to the config file.
+  The actual path for the config is "../deploy/configs/util/logger.json"
+  For the docker containers, the path is /etc/util/logger.json */
```
Comment (prasadtalasila) [inline util/logger.js]: please add a check about the existence of a proper value in "process.env.LOGGERCONFIG" variable. If the variable is not properly initialized, we can throw a meaningful error and bail out.

**Comment by AnkshitJain** [comment]
@prasadtalasila The requested changes have been incorporated. Please review the latest commit.

**Pull request merged**